File: java/src/com/v2soft/styxlib/library/messages/base/StyxMessage.java
package com.v2soft.styxlib.library.messages.base;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.nio.charset.Charset;

import com.v2soft.styxlib.library.io.IStyxDataReader;
import com.v2soft.styxlib.library.io.IStyxDataWriter;
import com.v2soft.styxlib.library.messages.StyxRAttachMessage;
import com.v2soft.styxlib.library.messages.StyxRAuthMessage;
import com.v2soft.styxlib.library.messages.StyxRErrorMessage;
import com.v2soft.styxlib.library.messages.StyxROpenMessage;
import com.v2soft.styxlib.library.messages.StyxRReadMessage;
import com.v2soft.styxlib.library.messages.StyxRStatMessage;
import com.v2soft.styxlib.library.messages.StyxRVersionMessage;
import com.v2soft.styxlib.library.messages.StyxRWalkMessage;
import com.v2soft.styxlib.library.messages.StyxRWriteMessage;
import com.v2soft.styxlib.library.messages.StyxTAttachMessage;
import com.v2soft.styxlib.library.messages.StyxTAuthMessage;
import com.v2soft.styxlib.library.messages.StyxTCreateMessage;
import com.v2soft.styxlib.library.messages.StyxTFlushMessage;
import com.v2soft.styxlib.library.messages.StyxTOpenMessage;
import com.v2soft.styxlib.library.messages.StyxTReadMessage;
import com.v2soft.styxlib.library.messages.StyxTVersionMessage;
import com.v2soft.styxlib.library.messages.StyxTWStatMessage;
import com.v2soft.styxlib.library.messages.StyxTWalkMessage;
import com.v2soft.styxlib.library.messages.StyxTWriteMessage;
import com.v2soft.styxlib.library.messages.base.enums.MessageType;
import com.v2soft.styxlib.library.messages.base.enums.ModeType;
import com.v2soft.styxlib.library.messages.base.structs.StyxQID;

public class StyxMessage {
    // ========================================================
    // Constants
    // ========================================================
    public static final Charset sUTFCharset = Charset.forName("utf-8"); 
    public static final int BASE_BINARY_SIZE = 7;

    public static final int NOTAG  =      0xFFFF;
    public static final long NOFID = 0xFFFFFFFFL;
    // ========================================================
    // Class fields
    // ========================================================
    private int mTag;
    private MessageType mType;

    /**
     * Construct message from DoubleStateBuffer
     * @param is intput stream
     * @param io_unit packet size
     * @return constructed Message object
     * @throws IOException
     */
    public static StyxMessage factory(IStyxDataReader buffer, int io_unit) 
            throws IOException {
        // get common packet data
        long packet_size = buffer.readUInt32();
        if ( packet_size > io_unit ) {
            throw new IOException("Packet size to large");
        }
        MessageType type = MessageType.factory(buffer.readUInt8());
        if ( type == null ) {
            throw new NullPointerException("Type is null, can't decode message");
        }
        int tag = buffer.readUInt16();
        // load other data
        StyxMessage result = null;
        switch (type) {
        case Tversion:
            result = new StyxTVersionMessage(0, null);
            break;
        case Rversion:
            result = new StyxRVersionMessage(0, null);
            break;
        case Tauth:
            result = new StyxTAuthMessage(NOFID);
            break;
        case Tflush:
            result = new StyxTFlushMessage(NOTAG);
            break;
        case Tattach:
            result = new StyxTAttachMessage(NOFID, NOFID, null, null);
            break;
        case Twalk:
            result = new StyxTWalkMessage(NOFID, NOFID, "");
            break;
        case Rauth:
            result = new StyxRAuthMessage(tag, StyxQID.EMPTY);
            break;
        case Rerror:
            result = new StyxRErrorMessage(tag, null);
            break;
        case Rflush:
            result = new StyxMessage(MessageType.Rflush, tag);
            break;
        case Rattach:
            result = new StyxRAttachMessage(tag, StyxQID.EMPTY);
            break;
        case Rwalk:
            result = new StyxRWalkMessage(tag, null); 
            break;
        case Topen:
            result = new StyxTOpenMessage(NOFID, ModeType.OREAD);
            break;
        case Ropen:
            result = new StyxROpenMessage(tag, null, 0, false);
            break;
        case Tcreate:
            result = new StyxTCreateMessage(NOFID, null, 0, ModeType.OWRITE);
            break;
        case Rcreate:
            result = new StyxROpenMessage(tag, null, 0, true);
            break;
        case Tread:
            result = new StyxTReadMessage(NOFID, null, 0);
            break;
        case Rread:
            result = new StyxRReadMessage(tag, null, 0);
            break;
        case Twrite:
            result = new StyxTWriteMessage(NOFID, null, null, 0, 0 );
            break;
        case Rwrite:
            result = new StyxRWriteMessage(tag, 0);
            break;
        case Tclunk:
            result = new StyxTMessageFID(MessageType.Tclunk, MessageType.Rclunk, 0);
            break;
        case Rclunk:
            result = new StyxMessage(MessageType.Rclunk, tag);
            break;
        case Tremove:
            result = new StyxTMessageFID(MessageType.Tremove, MessageType.Rremove, tag);
            break;
        case Rremove:
            result = new StyxMessage(MessageType.Rremove, tag);
            break;
        case Tstat:
            result = new StyxTMessageFID(MessageType.Tstat, MessageType.Rstat, tag);
            break;
        case Rstat:
            result = new StyxRStatMessage(tag);
            break;
        case Twstat:
            result = new StyxTWStatMessage(NOFID, null);
            break;
        case Rwstat:
            result = new StyxMessage(MessageType.Rwstat, tag);
            break;
        }
        result.setTag((short) tag);
        result.load(buffer);
        return result;
    }	

    public static String toString(byte[] bytes) {
        if ( (bytes == null) || (bytes.length==0))
            return "-";
        final StringBuilder result = new StringBuilder();
        result.append(Integer.toHexString(((int)bytes[0])&0xFF));
        result.append(",");
        int count = bytes.length;
        for (int i=1; i<count; i++)
        {
            result.append(Integer.toHexString(((int)bytes[i])&0xFF));
            result.append(',');
        }

        return String.format("(%s)", result);
    }

    public static int getUTFSize(String utf) {
        if (utf == null)
            return 2;
        return 2 + countUTFBytes(utf);
    }

    public static int countUTFBytes(String utf)	{
        String test = new String(utf);
        byte[] data = null;
        try {
            data = test.getBytes("utf-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        if ( data == null ) {
            return 0;
        } else {
            return data.length;
        }
    }

    public StyxMessage(MessageType type, int tag) {
        mType = type;
        mTag = tag;
    }

    public MessageType getType(){
        return mType;
    }

    public int getTag(){
        return mTag;
    }

    public void setTag(int tag){
        mTag = (tag & 0xFFFF);
    }

    public int getBinarySize(){
        return BASE_BINARY_SIZE;
    }

    protected void load(IStyxDataReader buffer) throws IOException {
    }
    public void writeToBuffer(IStyxDataWriter output)  
            throws UnsupportedEncodingException, IOException {
        output.clear();
        int packetSize = getBinarySize();
        output.limit(packetSize);
        output.writeUInt32(packetSize);
        output.writeUInt8((short) getType().getByte());
        output.writeUInt16(getTag());
    }

    @Override
    public String toString() {
        return String.format("Type: %s\nTag: %d", 
                getType().toString(), getTag());
    }
}

File: native/V2StyxLib/lib/messages/base/StyxMessage.cpp
/*
 * StyxMessage.cpp
 *
 *  Created on: May 24, 2012
 *      Author: vschryabets@gmail.com
 */

#include "stdio.h"
#include "types.h"
#include "messages/base/StyxMessage.h"
#include "messages/StyxTVersionMessage.h"
#include "messages/StyxRVersionMessage.h"
#include "messages/StyxTAttachMessage.h"
#include "messages/StyxTWalkMessage.h"
#include "messages/StyxTOpenMessage.h"
#include "messages/StyxTReadMessage.h"
#include "messages/StyxTFlushMessage.h"
#include "messages/StyxTWriteMessage.h"
#include "messages/StyxTWStatMessage.h"
#include "io/IStyxDataReader.h"
#include "messages/StyxTCreateMessage.h"
#include "messages/StyxRErrorMessage.h"
#include "messages/StyxRAttachMessage.h"
#include "messages/StyxRWalkMessage.h"
#include "messages/StyxROpenMessage.h"
#include "messages/StyxRReadMessage.h"
#include "messages/StyxRWriteMessage.h"
#include "messages/StyxRStatMessage.h"
#include <sstream>

size_t StyxMessage::getUTFSize(StyxString utf) {
	return utf.length();
}

StyxMessage::StyxMessage(MessageTypeEnum type, StyxTAG tag) {
	mType = type;
	mTag = tag;
}

StyxMessage::~StyxMessage() {
	// TODO Auto-generated destructor stub
}

/**
 * Construct message from DoubleStateBuffer
 * @param is intput stream
 * @param io_unit packet size
 * @return constructed Message object
 * @throws IOException
 */
StyxMessage* StyxMessage::factory(IStyxDataReader* buffer, size_t io_unit) {
	// get common packet data
	size_t packet_size = buffer->readUInt32();
	if ( packet_size > io_unit ) throw "Packet size to large";
	int type = buffer->readUInt8();
	uint16_t tag = buffer->readUInt16();
	// load other data
	StyxMessage *result = NULL;
	switch (type) {
	case Tversion:
		result = new StyxTVersionMessage(0, "");
		break;
	case Rversion:
		result = new StyxRVersionMessage(0, NULL);
		break;
		//	case Tauth:
		//		result = new StyxTAuthMessage(NOFID);
		//		break;
	case Tflush:
		result = new StyxTFlushMessage(NOTAG);
		break;
	case Tattach:
		result = new StyxTAttachMessage(NOFID, NOFID, "", "");
		break;
	case Twalk:
		result = new StyxTWalkMessage(NOFID, NOFID);
		break;
		//	case Rauth:
		//		result = new StyxRAuthMessage(tag, StyxQID.EMPTY);
		//		break;
    case Rerror:
        result = new StyxRErrorMessage(tag, NULL);
        break;
    case Rflush:
        result = new StyxMessage(Rflush, tag);
        break;
    case Rattach:
        result = new StyxRAttachMessage(tag, StyxQID::EMPTY);
        break;
    case Rwalk:
        result = new StyxRWalkMessage(tag, NULL);
        break;
	case Topen:
		result = new StyxTOpenMessage(NOFID, OREAD);
		break;
    case Ropen:
        result = new StyxROpenMessage(tag, NULL, 0, false);
        break;
    case Tcreate:
        result = new StyxTCreateMessage(NOFID, NULL, 0, OWRITE);
        break;
    case Rcreate:
        result = new StyxROpenMessage(tag, NULL, 0, true);
        break;
	case Tread:
		result = new StyxTReadMessage(NOFID, 0, 0);
		break;
    case Rread:
        result = new StyxRReadMessage(tag, NULL, 0);
        break;
	case Twrite:
		result = new StyxTWriteMessage(NOFID, 0, NULL, 0);
		break;
    case Rwrite:
        result = new StyxRWriteMessage(tag, 0);
        break;
	case Tclunk:
		result = new StyxTMessageFID(Tclunk, Rclunk, 0);
		break;
	case Rclunk:
		result = new StyxMessage(Rclunk, tag);
		break;
	case Tremove:
		result = new StyxTMessageFID(Tremove, Rremove, 0);
		break;
	case Rremove:
		result = new StyxMessage(Rremove, tag);
		break;
	case Tstat:
		result = new StyxTMessageFID(Tstat, Rstat, tag);
		break;
	case Rstat:
		result = new StyxRStatMessage(tag);
		break;
	case Twstat:
		result = new StyxTWStatMessage(NOFID, NULL, false);
		break;
	case Rwstat:
		result = new StyxMessage(Rwstat, tag);
		break;
	default:
		throw "Unknown message";
		return NULL;
	}
	result->setTag(tag);
	result->load(buffer);
	return result;
}

void StyxMessage::setTag(uint16_t tag) {
	mTag = tag;
}

MessageTypeEnum StyxMessage::getType() {
	return mType;
}

StyxTAG StyxMessage::getTag() {
	return mTag;
}

void StyxMessage::load(IStyxDataReader *buffer) {
}

void StyxMessage::writeToBuffer(IStyxDataWriter *output) {
	size_t packetSize = getBinarySize();
	output->clear();
	output->limit(packetSize);
	output->writeUInt32(packetSize);
	output->writeUInt8(mType);
	output->writeUInt16(getTag());
}

size_t StyxMessage::getBinarySize()	{
	return StyxMessage::BASE_BINARY_SIZE;
}

std::string StyxMessage::toString() {
	std::stringstream stream;
	stream << "Type " << getType() << " Tag " << getTag();
    throw stream.str();

}
